src/stenographer.c

```diff
@@ -0,0 +1,372 @@
+    curl_easy_setopt (st->curl, CURLOPT_SSLKEY, st->key);
+    curl_easy_setopt (st->curl, CURLOPT_CAINFO, st->cacert);
+    curl_easy_setopt (st->curl, CURLOPT_TCP_KEEPALIVE, 1L);
+
+    return (0);
+} /* }}} int init_host */
+
+static void submit_value (const char *type, const char *type_instance,
+        value_t value, stenographer_t *st)
+{
+    value_list_t vl = VALUE_LIST_INIT;
+
+    vl.values = &value;
+    vl.values_len = 1;
+
+    sstrncpy (vl.host, (st->host != NULL) ? st->host : hostname_g,
+            sizeof (vl.host));
+
+    sstrncpy (vl.plugin, "stenographer", sizeof (vl.plugin));
+    if (st->name != NULL)
+        sstrncpy (vl.plugin_instance, st->name,
+                sizeof (vl.plugin_instance));
+
+    sstrncpy (vl.type, type, sizeof (vl.type));
+    if (type_instance != NULL)
+        sstrncpy (vl.type_instance, type_instance,
+                sizeof (vl.type_instance));
+    plugin_dispatch_values (&vl);
+} /* void submit_value */
+
+static void submit_counter (const char *type, const char *type_instance,
+        counter_t c, stenographer_t *st)
+{
+    value_t v;
+    v.counter = c;
+    submit_value (type, type_instance, v, st);
+} /* void submit_counter */
+
+static int stenographer_read_host (user_data_t *user_data) /* {{{ */
+{
+    int  lines_num = 0;
+    int  curl_lines = 21; // number of steno stats in types.db
+    int  num_stat_fields = 2; // key with tab seperated  value 
+    char *line_ptr;
+    char *line_save_ptr;
+    char *stat_save_ptr;
+    char *lines[curl_lines];
+
+    stenographer_t *st;
+    st = user_data->data;
+    assert (st->url != NULL);
+
+    if (st->curl == NULL)
+    {
+        int status;
+
+        status = init_host (st);
+        if (status != 0)
+            return (-1);
+    }
+    
+    assert (st->curl != NULL);
+    st->stenographer_buffer_fill = 0;
+
+    // curl stats from stenographer server
+    if (curl_easy_perform (st->curl) != CURLE_OK)
+    {
+        ERROR ("stenographer: curl_easy_perform failed: %s",
+                st->stenographer_curl_error);
+        return (-1);
+    }
+
+    // parse stats
+    line_ptr = st->stenographer_buffer;
+    line_save_ptr = NULL;
+
+    while ((lines[lines_num] = strtok_r (line_ptr, "\n", &line_save_ptr)) != NULL)
+    {
+        line_ptr = NULL;
+        char *stat[num_stat_fields];
+        int stat_num = 0;
+        stat_save_ptr = NULL;
+        while((stat[stat_num] = strtok_r(lines[lines_num], "\t", &stat_save_ptr)) != NULL)
+        {
+           lines[lines_num] = NULL;
+
+           stat_num++;
+           if (stat_num >= num_stat_fields)
+               break;
+        }
+
+        char *prefix = "stenographer_";
+        int titleSize = strlen(prefix) + strlen(stat[0]) + 1;
+        char *title = (char *)malloc(titleSize);
+        if (title == NULL)
+        {
+            ERROR("stenographer plugin: prefix malloc failed.");
+            return (-1);
+        }
+        strcpy(title, prefix);
```

Comment:
Please use `ssnprintf (title, title_size, "%s%s", prefix, stat[0]);` instead of *strcpy(3)* / *strcat(3)*.